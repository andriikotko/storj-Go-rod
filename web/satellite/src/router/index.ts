// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vue from 'vue';
import Router, { RouteRecord } from 'vue-router';

import AccountArea from '@/components/account/AccountArea.vue';
import AccountBilling from '@/components/account/billing/BillingArea.vue';
import DetailedHistory from '@/components/account/billing/depositAndBillingHistory/DetailedHistory.vue';
import CreditsHistory from '@/components/account/billing/freeCredits/CreditsHistory.vue';
import SettingsArea from '@/components/account/SettingsArea.vue';
import ApiKeysArea from '@/components/apiKeys/ApiKeysArea.vue';
import Page404 from '@/components/errors/Page404.vue';
import OnboardingTourArea from '@/components/onboardingTour/OnboardingTourArea.vue';
import CreateProject from '@/components/project/CreateProject.vue';
import EditProjectDetails from '@/components/project/EditProjectDetails.vue';
import ProjectDashboard from '@/components/project/ProjectDashboard.vue';
import ProjectMembersArea from '@/components/team/ProjectMembersArea.vue';

import store from '@/store';
import { NavigationLink } from '@/types/navigation';
const DashboardArea = () => import('@/views/DashboardArea.vue');
const ForgotPassword = () => import('@/views/forgotPassword/ForgotPassword.vue');
const LoginArea = () => import('@/views/login/LoginArea.vue');
const RegisterArea = () => import('@/views/register/RegisterArea.vue');

Vue.use(Router);

/**
 * RouteConfig contains information about all routes and subroutes
 */
export abstract class RouteConfig {
    // root paths
    public static Root = new NavigationLink('/', 'Root');
    public static Login = new NavigationLink('/login', 'Login');
    public static Register = new NavigationLink('/signup', 'Register');
    public static ForgotPassword = new NavigationLink('/forgot-password', 'Forgot Password');
    public static Account = new NavigationLink('/account', 'Account');
    public static ProjectDashboard = new NavigationLink('/project-dashboard', 'Dashboard');
    public static Users = new NavigationLink('/project-members', 'Users');
    public static ApiKeys = new NavigationLink('/api-keys', 'API Keys');
    public static OnboardingTour = new NavigationLink('/onboarding-tour', 'Onboarding Tour');
    public static CreateProject = new NavigationLink('/create-project', 'Create Project');
    public static EditProjectDetails = new NavigationLink('/edit-project-details', 'Edit Project Details');

    // child paths
    public static Settings = new NavigationLink('settings', 'Settings');
    public static Billing = new NavigationLink('billing', 'Billing');
    public static BillingHistory = new NavigationLink('billing-history', 'Billing History');
    public static DepositHistory = new NavigationLink('deposit-history', 'Deposit History');
    public static CreditsHistory = new NavigationLink('credits-history', 'Credits History');
    // TODO: disabled until implementation
    // public static Referral = new NavigationLink('referral', 'Referral');

    // not in project yet
    // public static Referral = new NavigationLink('//ref/:ids', 'Referral');
}

export const notProjectRelatedRoutes = [
    RouteConfig.Login.name,
    RouteConfig.Register.name,
    RouteConfig.ForgotPassword.name,
    RouteConfig.Billing.name,
    RouteConfig.BillingHistory.name,
    RouteConfig.DepositHistory.name,
    RouteConfig.CreditsHistory.name,
    RouteConfig.Settings.name,
    // RouteConfig.Referral.name,
];

export const router = new Router({
    mode: 'history',
    routes: [
        {
            path: RouteConfig.Login.path,
            name: RouteConfig.Login.name,
            component: LoginArea,
        },
        {
            path: RouteConfig.Register.path,
            name: RouteConfig.Register.name,
            component: RegisterArea,
        },
        {
            path: RouteConfig.ForgotPassword.path,
            name: RouteConfig.ForgotPassword.name,
            component: ForgotPassword,
        },
        {
            path: RouteConfig.Root.path,
            meta: {
                requiresAuth: true,
            },
            component: DashboardArea,
            children: [
                {
                    path: RouteConfig.Account.path,
                    name: RouteConfig.Account.name,
                    component: AccountArea,
                    children: [
                        {
                            path: RouteConfig.Settings.path,
                            name: RouteConfig.Settings.name,
                            component: SettingsArea,
                        },
                        {
                            path: RouteConfig.Billing.path,
                            name: RouteConfig.Billing.name,
                            component: AccountBilling,
                        },
                        {
                            path: RouteConfig.BillingHistory.path,
                            name: RouteConfig.BillingHistory.name,
                            component: DetailedHistory,
                        },
                        {
                            path: RouteConfig.DepositHistory.path,
                            name: RouteConfig.DepositHistory.name,
                            component: DetailedHistory,
                        },
                        {
                            path: RouteConfig.CreditsHistory.path,
                            name: RouteConfig.CreditsHistory.name,
                            component: CreditsHistory,
                        },
                        // {
                        //     path: RouteConfig.Referral.path,
                        //     name: RouteConfig.Referral.name,
                        //     component: ReferralArea,
                        // },
                    ],
                },
                {
                    path: RouteConfig.ProjectDashboard.path,
                    name: RouteConfig.ProjectDashboard.name,
                    component: ProjectDashboard,
                },
                {
                    path: RouteConfig.Root.path,
                    name: 'default',
                    component: ProjectDashboard,
                },
                {
                    path: RouteConfig.Users.path,
                    name: RouteConfig.Users.name,
                    component: ProjectMembersArea,
                },
                {
                    path: RouteConfig.ApiKeys.path,
                    name: RouteConfig.ApiKeys.name,
                    component: ApiKeysArea,
                },
                {
                    path: RouteConfig.OnboardingTour.path,
                    name: RouteConfig.OnboardingTour.name,
                    component: OnboardingTourArea,
                },
                {
                    path: RouteConfig.CreateProject.path,
                    name: RouteConfig.CreateProject.name,
                    component: CreateProject,
                },
                {
                    path: RouteConfig.EditProjectDetails.path,
                    name: RouteConfig.EditProjectDetails.name,
                    component: EditProjectDetails,
                },
            ],
        },
        {
            path: '*',
            name: '404',
            component: Page404,
        },
    ],
});

router.beforeEach((to, from, next) => {
    if (navigateToDefaultSubTab(to.matched, RouteConfig.Account)) {
        next(RouteConfig.Account.with(RouteConfig.Billing).path);

        return;
    }

    if (to.name === 'default') {
        next(RouteConfig.ProjectDashboard.path);

        return;
    }

    next();
});

router.afterEach(({name}, from) => {
    if (!name) {
        return;
    }

    if (notProjectRelatedRoutes.includes(name)) {
        document.title = `${router.currentRoute.name} | ${store.state.appStateModule.satelliteName}`;

        return;
    }

    const selectedProjectName = store.state.projectsModule.selectedProject.name ?
        `${store.state.projectsModule.selectedProject.name} | ` : '';

    document.title = `${selectedProjectName + router.currentRoute.name} | ${store.state.appStateModule.satelliteName}`;
});

/**
 * if our route is a tab and has no sub tab route - we will navigate to default subtab.
 * F.E. /account/ -> /account/billing/;
 * @param routes - array of RouteRecord from vue-router
 * @param next - callback to process next route
 * @param tabRoute - tabNavigator route
 */
function navigateToDefaultSubTab(routes: RouteRecord[], tabRoute: NavigationLink): boolean {
    return routes.length === 2 && (routes[1].name as string) === tabRoute.name;
}
