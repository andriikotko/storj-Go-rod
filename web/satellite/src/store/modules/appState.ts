// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';
import { AppState } from '@/utils/constants/appStateEnum';

import { APP_STATE_MUTATIONS } from '../mutationConstants';

export const appStateModule = {
    state: {
        // Object that contains all states of views
        appState: {
            fetchState: AppState.LOADING,
            isAddTeamMembersPopupShown: false,
            isAccountDropdownShown: false,
            isSelectProjectDropdownShown: false,
            isResourcesDropdownShown: false,
            isSettingsDropdownShown: false,
            isEditProjectDropdownShown: false,
            isFreeCreditsDropdownShown: false,
            isAvailableBalanceDropdownShown: false,
            isPeriodsDropdownShown: false,
            isDeleteProjectPopupShown: false,
            isDeleteAccountPopupShown: false,
            isSortProjectMembersByPopupShown: false,
            isSuccessfulRegistrationShown: false,
            isEditProfilePopupShown: false,
            isChangePasswordPopupShown: false,
            isPaymentSelectionShown: false,
            isCreateProjectButtonShown: false,
            isSaveApiKeyModalShown: false,
        },
        satelliteName: '',
    },
    mutations: {
        // Mutation changing add projectMembers members popup visibility
        [APP_STATE_MUTATIONS.TOGGLE_ADD_TEAMMEMBER_POPUP](state: any): void {
            state.appState.isAddTeamMembersPopupShown = !state.appState.isAddTeamMembersPopupShown;
        },
        // Mutation changing save api key modal visibility
        [APP_STATE_MUTATIONS.TOGGLE_SAVE_API_KEY_MODAL](state: any): void {
            state.appState.isSaveApiKeyModalShown = !state.appState.isSaveApiKeyModalShown;
        },
        // Mutation changing account dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_ACCOUNT_DROPDOWN](state: any): void {
            state.appState.isAccountDropdownShown = !state.appState.isAccountDropdownShown;
        },
        // Mutation changing select project dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_SELECT_PROJECT_DROPDOWN](state: any): void {
            state.appState.isSelectProjectDropdownShown = !state.appState.isSelectProjectDropdownShown;
        },
        // Mutation changing resources dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_RESOURCES_DROPDOWN](state: any): void {
            state.appState.isResourcesDropdownShown = !state.appState.isResourcesDropdownShown;
        },
        // Mutation changing settings dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_SETTINGS_DROPDOWN](state: any): void {
            state.appState.isSettingsDropdownShown = !state.appState.isSettingsDropdownShown;
        },
        // Mutation changing edit project dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_EDIT_PROJECT_DROPDOWN](state: any): void {
            state.appState.isEditProjectDropdownShown = !state.appState.isEditProjectDropdownShown;
        },
        // Mutation changing free credits dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_FREE_CREDITS_DROPDOWN](state: any): void {
            state.appState.isFreeCreditsDropdownShown = !state.appState.isFreeCreditsDropdownShown;
        },
        // Mutation changing available balance dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_AVAILABLE_BALANCE_DROPDOWN](state: any): void {
            state.appState.isAvailableBalanceDropdownShown = !state.appState.isAvailableBalanceDropdownShown;
        },
        // Mutation changing periods dropdown visibility
        [APP_STATE_MUTATIONS.TOGGLE_PERIODS_DROPDOWN](state: any): void {
            state.appState.isPeriodsDropdownShown = !state.appState.isPeriodsDropdownShown;
        },
        // Mutation changing delete project popup visibility
        [APP_STATE_MUTATIONS.TOGGLE_DELETE_PROJECT_DROPDOWN](state: any): void {
            state.appState.isDeleteProjectPopupShown = !state.appState.isDeleteProjectPopupShown;
        },
        // Mutation changing delete account popup visibility
        [APP_STATE_MUTATIONS.TOGGLE_DELETE_ACCOUNT_DROPDOWN](state: any): void {
            state.appState.isDeleteAccountPopupShown = !state.appState.isDeleteAccountPopupShown;
        },
        // Mutation changing 'sort project members by' popup visibility.
        [APP_STATE_MUTATIONS.TOGGLE_SORT_PM_BY_DROPDOWN](state: any): void {
            state.appState.isSortProjectMembersByPopupShown = !state.appState.isSortProjectMembersByPopupShown;
        },
        // Mutation changing 'successful registration' area visibility.
        [APP_STATE_MUTATIONS.TOGGLE_SUCCESSFUL_REGISTRATION](state: any): void {
            state.appState.isSuccessfulRegistrationShown = !state.appState.isSuccessfulRegistrationShown;
        },
        [APP_STATE_MUTATIONS.TOGGLE_CHANGE_PASSWORD_POPUP](state: any): void {
            state.appState.isChangePasswordPopupShown = !state.appState.isChangePasswordPopupShown;
        },
        [APP_STATE_MUTATIONS.TOGGLE_EDIT_PROFILE_POPUP](state: any): void {
            state.appState.isEditProfilePopupShown = !state.appState.isEditProfilePopupShown;
        },
        [APP_STATE_MUTATIONS.SHOW_SET_DEFAULT_PAYMENT_METHOD_POPUP](state: any, id: string): void {
            state.appState.setDefaultPaymentMethodID = id;
        },
        [APP_STATE_MUTATIONS.SHOW_DELETE_PAYMENT_METHOD_POPUP](state: any, id: string): void {
            state.appState.deletePaymentMethodID = id;
        },
        [APP_STATE_MUTATIONS.SHOW_CREATE_PROJECT_BUTTON](state: any): void {
            state.appState.isCreateProjectButtonShown = true;
        },
        [APP_STATE_MUTATIONS.HIDE_CREATE_PROJECT_BUTTON](state: any): void {
            state.appState.isCreateProjectButtonShown = false;
        },
        // Mutation that closes each popup/dropdown
        [APP_STATE_MUTATIONS.CLOSE_ALL](state: any): void {
            state.appState.isAccountDropdownShown = false;
            state.appState.isSelectProjectDropdownShown = false;
            state.appState.isResourcesDropdownShown = false;
            state.appState.isSettingsDropdownShown = false;
            state.appState.isEditProjectDropdownShown = false;
            state.appState.isFreeCreditsDropdownShown = false;
            state.appState.isAvailableBalanceDropdownShown = false;
            state.appState.isPeriodsDropdownShown = false;
            state.appState.isPaymentSelectionShown = false;
            state.appState.isSortProjectMembersByPopupShown = false;
        },
        [APP_STATE_MUTATIONS.CHANGE_STATE](state: any, newFetchState: AppState): void {
            state.appState.fetchState = newFetchState;
        },
        // Mutation changing payment selection visibility
        [APP_STATE_MUTATIONS.TOGGLE_PAYMENT_SELECTION](state: any, value: boolean): void {
            state.appState.isPaymentSelectionShown = value;
        },
        [APP_STATE_MUTATIONS.SET_NAME](state: any, satelliteName: string): void {
            state.satelliteName = satelliteName;
        },
    },
    actions: {
        // Commits mutation for changing app popups and dropdowns visibility state
        [APP_STATE_ACTIONS.TOGGLE_TEAM_MEMBERS]: function ({commit, state}: any): void {
            if (!state.appState.isAddTeamMembersPopupShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_ADD_TEAMMEMBER_POPUP);
        },
        [APP_STATE_ACTIONS.TOGGLE_SAVE_API_KEY_MODAL]: function ({commit, state}: any): void {
            if (!state.appState.isSaveApiKeyModalShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_SAVE_API_KEY_MODAL);
        },
        [APP_STATE_ACTIONS.TOGGLE_ACCOUNT]: function ({commit, state}: any): void {
            if (!state.appState.isAccountDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_ACCOUNT_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_SELECT_PROJECT_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isSelectProjectDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_SELECT_PROJECT_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_RESOURCES_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isResourcesDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_RESOURCES_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_SETTINGS_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isSettingsDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_SETTINGS_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_EDIT_PROJECT_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isEditProjectDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_EDIT_PROJECT_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_FREE_CREDITS_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isFreeCreditsDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_FREE_CREDITS_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_AVAILABLE_BALANCE_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isAvailableBalanceDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_AVAILABLE_BALANCE_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_PERIODS_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isPeriodsDropdownShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_PERIODS_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_PAYMENT_SELECTION]: function ({commit, state}: any, value: boolean): void {
            if (!state.appState.isPaymentSelectionShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_PAYMENT_SELECTION, value);
        },
        [APP_STATE_ACTIONS.TOGGLE_DEL_PROJ]: function ({commit, state}: any): void {
            if (!state.appState.isDeleteProjectPopupShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_DELETE_PROJECT_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_DEL_ACCOUNT]: function ({commit, state}: any): void {
            if (!state.appState.isDeleteAccountPopupShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_DELETE_ACCOUNT_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_SORT_PM_BY_DROPDOWN]: function ({commit, state}: any): void {
            if (!state.appState.isSortProjectMembersByPopupShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_SORT_PM_BY_DROPDOWN);
        },
        [APP_STATE_ACTIONS.TOGGLE_SUCCESSFUL_REGISTRATION]: function ({commit, state}: any): void {
            if (!state.appState.isSuccessfulRegistrationShown) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.TOGGLE_SUCCESSFUL_REGISTRATION);
        },
        [APP_STATE_ACTIONS.TOGGLE_CHANGE_PASSWORD_POPUP]: function ({commit}: any): void {
            commit(APP_STATE_MUTATIONS.TOGGLE_CHANGE_PASSWORD_POPUP);
        },
        [APP_STATE_ACTIONS.TOGGLE_EDIT_PROFILE_POPUP]: function ({commit}: any): void {
            commit(APP_STATE_MUTATIONS.TOGGLE_EDIT_PROFILE_POPUP);
        },
        [APP_STATE_ACTIONS.SHOW_SET_DEFAULT_PAYMENT_METHOD_POPUP]: function ({commit, state}: any, methodID: string): void {
            if (!state.appState.setDefaultPaymentMethodID) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.SHOW_SET_DEFAULT_PAYMENT_METHOD_POPUP, methodID);
        },
        [APP_STATE_ACTIONS.SHOW_DELETE_PAYMENT_METHOD_POPUP]: function ({commit, state}: any, methodID: string): void {
            if (!state.appState.deletePaymentMethodID) {
                commit(APP_STATE_MUTATIONS.CLOSE_ALL);
            }

            commit(APP_STATE_MUTATIONS.SHOW_DELETE_PAYMENT_METHOD_POPUP, methodID);
        },
        [APP_STATE_ACTIONS.SHOW_CREATE_PROJECT_BUTTON]: function ({commit}: any): void {
            commit(APP_STATE_MUTATIONS.SHOW_CREATE_PROJECT_BUTTON);
        },
        [APP_STATE_ACTIONS.HIDE_CREATE_PROJECT_BUTTON]: function ({commit}: any): void {
            commit(APP_STATE_MUTATIONS.HIDE_CREATE_PROJECT_BUTTON);
        },
        [APP_STATE_ACTIONS.CLOSE_POPUPS]: function ({commit}: any): void {
            commit(APP_STATE_MUTATIONS.CLOSE_ALL);
        },
        [APP_STATE_ACTIONS.CHANGE_STATE]: function ({commit}: any, newFetchState: AppState): void {
            commit(APP_STATE_MUTATIONS.CHANGE_STATE, newFetchState);
        },
        [APP_STATE_ACTIONS.SET_SATELLITE_NAME]: function ({commit}: any, satelliteName: string): void {
            commit(APP_STATE_MUTATIONS.SET_NAME, satelliteName);
        },
    },
};
