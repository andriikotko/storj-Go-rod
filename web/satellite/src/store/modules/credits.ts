// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { CreditsApi, CreditUsage } from '@/types/credits';

export const CREDIT_USAGE_ACTIONS = {
    FETCH: 'fetchCreditUsage',
    CLEAR: 'clearCreditUsage',
};

export const CREDIT_USAGE_MUTATIONS = {
    SET: 'SET_CREDIT_USAGE',
    CLEAR: 'CLEAR_CREDIT_USAGE',
};

const { FETCH } = CREDIT_USAGE_ACTIONS;
const { SET, CLEAR } = CREDIT_USAGE_MUTATIONS;

export function makeCreditsModule(api: CreditsApi): StoreModule<CreditUsage> {
    return {
        state: new CreditUsage(),

        mutations: {
            [SET](state: CreditUsage, creditUsage: CreditUsage) {
                state.availableCredits = creditUsage.availableCredits;
                state.usedCredits = creditUsage.usedCredits;
                state.referred = creditUsage.referred;
            },
            [CLEAR](state: CreditUsage): void {
                state.availableCredits = 0;
                state.usedCredits = 0;
                state.referred = 0;
            },
        },

        actions: {
            [FETCH]: async function({commit}: any): Promise<CreditUsage> {
                const credits = await api.get();

                commit(SET, credits);

                return credits;
            },
            [CREDIT_USAGE_ACTIONS.CLEAR]: function({commit}: any): void {
                commit(CLEAR);
            },
        },

        getters: {
            credits: (state: CreditUsage): CreditUsage => state,
        },
    };
}
