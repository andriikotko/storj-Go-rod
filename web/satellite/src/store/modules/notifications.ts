// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { DelayedNotification } from '@/types/DelayedNotification';
import { NOTIFICATION_ACTIONS } from '@/utils/constants/actionNames';
import { NOTIFICATION_TYPES } from '@/utils/constants/notification';

import { NOTIFICATION_MUTATIONS } from '../mutationConstants';

export class NotificationsState {
    public notificationQueue: DelayedNotification[] = [];
}

export function makeNotificationsModule(): StoreModule<NotificationsState> {
    return {
        state: new NotificationsState(),
        mutations: {
            // Mutation for adding notification to queue
            [NOTIFICATION_MUTATIONS.ADD](state: any, notification: DelayedNotification): void {
                state.notificationQueue.push(notification);
            },
            // Mutation for deleting notification to queue
            [NOTIFICATION_MUTATIONS.DELETE](state: any, id: string): void {
                if (state.notificationQueue.length < 1) {
                    return;
                }

                const selectedNotification =  getNotificationById(state.notificationQueue, id);
                if (selectedNotification) {
                    selectedNotification.pause();
                    state.notificationQueue.splice(state.notificationQueue.indexOf(selectedNotification), 1);
                }
            },
            [NOTIFICATION_MUTATIONS.PAUSE](state: any, id: string): void {
                const selectedNotification =  getNotificationById(state.notificationQueue, id);
                if (selectedNotification) {
                    selectedNotification.pause();
                }
            },
            [NOTIFICATION_MUTATIONS.RESUME](state: any, id: string): void {
                const selectedNotification =  getNotificationById(state.notificationQueue, id);
                if (selectedNotification) {
                    selectedNotification.start();
                }
            },
            [NOTIFICATION_MUTATIONS.CLEAR](state: any): void {
                state.notificationQueue = [];
            },
        },
        actions: {
            // Commits mutation for adding success notification
            [NOTIFICATION_ACTIONS.SUCCESS]: function ({commit}: any, message: string): void {
                const notification = new DelayedNotification(
                    () => commit(NOTIFICATION_MUTATIONS.DELETE, notification.id),
                    NOTIFICATION_TYPES.SUCCESS,
                    message,
                );

                commit(NOTIFICATION_MUTATIONS.ADD, notification);
            },
            // Commits mutation for adding info notification
            [NOTIFICATION_ACTIONS.NOTIFY]: function ({commit}: any, message: string): void {

                const notification = new DelayedNotification(
                    () => commit(NOTIFICATION_MUTATIONS.DELETE, notification.id),
                    NOTIFICATION_TYPES.NOTIFICATION,
                    message,
                );

                commit(NOTIFICATION_MUTATIONS.ADD, notification);
            },
            // Commits mutation for adding error notification
            [NOTIFICATION_ACTIONS.ERROR]: function ({commit}: any, message: string): void {
                const notification = new DelayedNotification(
                    () => commit(NOTIFICATION_MUTATIONS.DELETE, notification.id),
                    NOTIFICATION_TYPES.ERROR,
                    message,
                );

                commit(NOTIFICATION_MUTATIONS.ADD, notification);
            },
            // Commits mutation for adding error notification
            [NOTIFICATION_ACTIONS.WARNING]: function ({commit}: any, message: string): void {
                const notification = new DelayedNotification(
                    () => commit(NOTIFICATION_MUTATIONS.DELETE, notification.id),
                    NOTIFICATION_TYPES.WARNING,
                    message,
                );

                commit(NOTIFICATION_MUTATIONS.ADD, notification);
            },
            [NOTIFICATION_ACTIONS.DELETE]: function ({commit}: any, id: string): void {
                commit(NOTIFICATION_MUTATIONS.DELETE, id);
            },
            [NOTIFICATION_ACTIONS.PAUSE]: function ({commit}: any, id: string): void {
                commit(NOTIFICATION_MUTATIONS.PAUSE, id);
            },
            [NOTIFICATION_ACTIONS.RESUME]: function ({commit}: any, id: string): void {
                commit(NOTIFICATION_MUTATIONS.RESUME, id);
            },
            [NOTIFICATION_ACTIONS.CLEAR]: function ({commit}): void {
                commit(NOTIFICATION_MUTATIONS.CLEAR);
            },
        },
        getters: {},
    };
}

function getNotificationById(notifications: DelayedNotification[], id: string): DelayedNotification | undefined {
    return notifications.find((notification: DelayedNotification) => notification.id === id);
}
