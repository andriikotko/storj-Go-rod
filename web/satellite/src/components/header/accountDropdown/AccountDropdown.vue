// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="account-dropdown">
        <div class="account-dropdown__wrap">
            <div class="account-dropdown__wrap__name-container">
                <p class="account-dropdown__wrap__name-container__name">{{ userName }}</p>
            </div>
            <div class="account-dropdown__wrap__item-container" @click.stop="onLogoutClick">
                <p class="account-dropdown__wrap__item-container__title">Sign Out</p>
                <p class="account-dropdown__wrap__item-container__arrow">-></p>
            </div>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import { AuthHttpApi } from '@/api/auth';
import { RouteConfig } from '@/router';
import { API_KEYS_ACTIONS } from '@/store/modules/apiKeys';
import { BUCKET_ACTIONS } from '@/store/modules/buckets';
import { PROJECTS_ACTIONS } from '@/store/modules/projects';
import { USER_ACTIONS } from '@/store/modules/users';
import {
    APP_STATE_ACTIONS,
    NOTIFICATION_ACTIONS,
    PM_ACTIONS,
} from '@/utils/constants/actionNames';
import { LocalData } from '@/utils/localData';

@Component
export default class AccountDropdown extends Vue {
    private readonly auth: AuthHttpApi = new AuthHttpApi();

    /**
     * Returns user's full name.
     */
    public get userName(): string {
        return this.$store.getters.userName;
    }

    /**
     * Performs logout on backend than clears all user information from store and local storage.
     */
    public async onLogoutClick(): Promise<void> {
        try {
            await this.auth.logout();
        } catch (error) {
            await this.$notify.error(error.message);

            return;
        }

        await this.$router.push(RouteConfig.Login.path);
        await this.$store.dispatch(PM_ACTIONS.CLEAR);
        await this.$store.dispatch(PROJECTS_ACTIONS.CLEAR);
        await this.$store.dispatch(USER_ACTIONS.CLEAR);
        await this.$store.dispatch(API_KEYS_ACTIONS.CLEAR);
        await this.$store.dispatch(NOTIFICATION_ACTIONS.CLEAR);
        await this.$store.dispatch(BUCKET_ACTIONS.CLEAR);
        await this.$store.dispatch(APP_STATE_ACTIONS.CLOSE_POPUPS);

        LocalData.removeUserId();
    }
}
</script>

<style scoped lang="scss">
    .account-dropdown {
        position: absolute;
        top: 65px;
        right: 0;
        padding: 6px 0;
        background-color: #f5f6fa;
        z-index: 1120;
        box-shadow: 0 20px 34px 0 rgba(10, 27, 44, 0.28);
        border-radius: 6px;

        &__wrap {
            font-family: 'font_regular', sans-serif;
            min-width: 250px;
            max-width: 250px;
            background-color: #f5f6fa;
            border-radius: 6px;

            &__name-container {
                width: calc(100% - 45px);
                margin-left: 30px;
                padding: 0 15px 15px 0;

                &__name {
                    font-size: 14px;
                    line-height: 19px;
                    color: #1b2533;
                    white-space: nowrap;
                    overflow: hidden;
                    text-overflow: ellipsis;
                    margin: 15px 0 0 0;
                }
            }

            &__item-container {
                display: flex;
                align-items: center;
                justify-content: space-between;
                padding: 20px 20px 20px 30px;
                cursor: pointer;

                &__title,
                &__arrow {
                    font-size: 14px;
                    line-height: 19px;
                    color: #0068dc;
                    font-weight: 500;
                    margin: 0;
                }
            }
        }
    }
</style>
