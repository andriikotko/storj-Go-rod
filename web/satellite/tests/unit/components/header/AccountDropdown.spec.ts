// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

import sinon from 'sinon';
import Vuex from 'vuex';

import AccountDropdown from '@/components/header/accountDropdown/AccountDropdown.vue';

import { RouteConfig, router } from '@/router';
import { appStateModule } from '@/store/modules/appState';
import { createLocalVue, mount } from '@vue/test-utils';

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store({ modules: { appStateModule }});

describe('AccountDropdown', () => {
    it('renders correctly', (): void => {
        const wrapper = mount(AccountDropdown, {
            store,
            localVue,
            router,
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('router works correctly', async (): Promise<void> => {
        const routerSpy = sinon.spy();
        const wrapper = mount(AccountDropdown, {
            store,
            localVue,
            router,
        });

        wrapper.vm.onLogoutClick = routerSpy;

        await wrapper.find('.account-dropdown__wrap__item-container').trigger('click');

        expect(routerSpy.callCount).toBe(1);
    });
});
