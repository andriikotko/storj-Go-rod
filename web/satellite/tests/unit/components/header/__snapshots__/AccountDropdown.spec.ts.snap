// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDropdown renders correctly 1`] = `
<div class="account-dropdown">
  <div class="account-dropdown__wrap">
    <div class="account-dropdown__wrap__name-container">
      <p class="account-dropdown__wrap__name-container__name"></p>
    </div>
    <div class="account-dropdown__wrap__item-container">
      <p class="account-dropdown__wrap__item-container__title">Sign Out</p>
      <p class="account-dropdown__wrap__item-container__arrow">-&gt;</p>
    </div>
  </div>
</div>
`;
